File: Algorithms/C++/Medium/105_Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        if (preorder.empty() || inorder.empty() || preorder.size() != inorder.size()) {
            return NULL;
        }

        return helper(preorder, 0, preorder.size() - 1, inorder, 0, inorder.size() - 1);
    }
private:
    TreeNode* helper(vector<int>& preorder, int preorder_start, int preorder_end, vector<int>& inorder, int inorder_start, int inorder_end) {
        if (inorder_start > inorder_end) {
            return NULL;
        }

        int root_val = preorder[preorder_start];
        TreeNode* root = new TreeNode(root_val);

        int inorder_root_index;
        for (inorder_root_index = inorder_start; inorder_root_index <= inorder_end && inorder[inorder_root_index] != root_val; inorder_root_index++) {
            ;
        }
        int len = inorder_root_index - inorder_start;

        root->left = helper(preorder, preorder_start + 1, preorder_start + 1 + len - 1, inorder, inorder_start, inorder_root_index - 1);
        root->right = helper(preorder, preorder_start + 1 + len - 1 + 1, preorder_end, inorder, inorder_root_index + 1, inorder_end);

        return root;
    }
};

File: Algorithms/Swift/Medium/105_Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.left = nil
 *         self.right = nil
 *     }
 * }
 */
class Solution {
    func buildTree(preorder: [Int], _ inorder: [Int]) -> TreeNode? {
        if preorder.isEmpty || inorder.isEmpty || preorder.count != inorder.count {
            return nil
        }

        return helper(preorder: preorder, preorderStart: 0, preorderEnd: preorder.count - 1, inorder: inorder, inorderStart: 0, inorderEnd: inorder.count - 1)
    }

    private func helper(preorder preorder: [Int], preorderStart: Int, preorderEnd: Int, inorder: [Int], inorderStart: Int, inorderEnd: Int) -> TreeNode? {
        if preorderStart > preorderEnd {
            return nil
        }

        let rootVal = preorder[preorderStart]
        let root = TreeNode(rootVal)

        var inorderIndex = 0
        for i in inorderStart...inorderEnd {
            if inorder[i] == rootVal {
                inorderIndex = i
                print(inorderIndex)
                break
            }
        }

        let length = inorderIndex - inorderStart

        root.left = helper(preorder: preorder, preorderStart: preorderStart + 1, preorderEnd: preorderStart + 1 + length - 1, inorder: inorder, inorderStart: inorderStart, inorderEnd: inorderIndex - 1)
        root.right = helper(preorder: preorder, preorderStart: preorderStart + 1 + length - 1 + 1, preorderEnd: preorderEnd, inorder: inorder, inorderStart: inorderIndex + 1, inorderEnd: inorderEnd)

        return root
    }
}
